src/lib/Occurrence.hs

```diff
@@ -0,0 +1,623 @@
+  -- from multiple references).
+  -- Corresponds to sequential statements (e.g., x.i; x.j)
+  All :: [Access n] -> Access n
+  -- The object may be accessed in any of the ways given by the second
+  -- argument, as the name in the `Binder` argument varies over its
+  -- type.
+  -- Corresponds to `for`.
+  ForEach :: forall n l. (NameBinder AtomNameC n l) -> Access l -> Access n
+
+-- These are the indexing expressions we model.
+data IxExpr (n::S) =
+  -- Indexing by an identifier in scope
+    Var (Name AtomNameC n)
+  -- An index of product type
+  | Product [IxExpr n]
+  -- An index of sum type
+  | Inject Int (IxExpr n)
+  -- Unknown function of the given names
+  | Unknown [Name AtomNameC n]
+  -- All indices are accessed.  This constructor only appears after
+  -- `iterate`, below.
+  | IxAll
+  -- TODO Could add more constructors to capture other indexing
+  -- expressions, such as `from_ordinal _ $ ordinal i + 1`.  If that's
+  -- the only indexing expression the binding is safe to inline, and
+  -- inlining even preserves locality, not just work.
+  deriving (Eq, Show)
+
+-- We give Access semantics by a notional translation into a piece of
+-- code that dynamically counts the number of times every index is
+-- actually accessed.
+--
+-- - An Access becomes a function consuming a buffer reference
+--   and returning unit.  The buffer has the same type as the
+--   object whose accesses we are analyzing, except for integer
+--   leaves.
+--
+-- - For every control path of the source program, there is a control
+--   path of the Access program that accumulates at least as many
+--   counts into each element of the buffer as the number of times the
+--   source program read that element.
+--
+-- - `Location exprs ct` turns into accumulation
+--     \ref. ref!<exprs> += ct
+--     Note that this += must be understood as pointwise if
+--     the source indexing expression is not fully saturated.
+--
+-- - `Any opts` turns into case
+--     \ref. index = randint $ length opts
+--       case index of
+--         1 -> compile opt1 ref
+--         2 -> compile opt2 ref
+--         ...
+--
+-- - `All parts` turns into sequentialization
+--     \ref.
+--       compile part1 ref
+--       compile part2 ref
+--       ...
+--
+-- - `ForEach b sub` is iteration
+--     \ref. for b in (type of b).
+--       compile sub ref
+
+-- === Use ===
+
+-- The Use structure is a "data space" representation of accessing
+-- arrays.
+--
+-- Note that a Use can still depend on iteration-space indices (from
+-- the `iter` scope)---it just doesn't have internal binders that are
+-- meant to be interpreted in iteration space.
+--
+-- A Use implicitly iterates over as many dimensions of the accessed
+-- object as there are nesting levels of At constructors in the Use.
+-- We do not use explicit binders for this because they always appear
+-- in the same nesting order and can always be pulled to the outermost
+-- layer, so it's simplest to leave them implicit.
```

Comment:
I don't understand this comment and the previous one

Reply:
Tried to clarify by reordering text and giving examples.